cpp/tests/bitmask/bit_mask_test.cu

```diff
@@ -0,0 +1,436 @@
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 1));
+
+  BitMask bit_mask(bits, num_rows);
+
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+
+  EXPECT_EQ(10U, local_count);
+}
+
+TEST_F(BitMaskTest, MultipleOfEight)
+{
+  const int num_rows = 1024;
+
+  bit_mask_t *bits = nullptr;
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 0));
+
+  BitMask bit_mask(bits, num_rows);
+
+  for (int i = 0 ; i < num_rows ; i += 8) {
+    set_bit<<<1,1>>>(i, bit_mask);
+  }
+  
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+
+  EXPECT_EQ(128U, local_count);
+}
+
+TEST_F(BitMaskTest, NotMultipleOfEight)
+{
+  const int num_rows = 1023;
+
+  bit_mask_t *bits = nullptr;
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 0));
+
+  BitMask bit_mask(bits, num_rows);
+
+  for (int i = 7 ; i < num_rows ; i += 8) {
+    set_bit<<<1,1>>>(i, bit_mask);
+  }
+  
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+
+  EXPECT_EQ(127U, local_count);
+}
+
+TEST_F(BitMaskTest, TenThousandRows)
+{
+  const int num_rows = 10000;
+
+  bit_mask_t *bits = nullptr;
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 1));
+
+  BitMask bit_mask(bits, num_rows);
+
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+
+  EXPECT_EQ(10000U, local_count);
+}
+
+TEST_F(BitMaskTest, PerformanceTest)
+{
+  const int num_rows = 100000000;
+
+  bit_mask_t *bits = nullptr;
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 0));
+
+  BitMask bit_mask(bits, num_rows);
+
+  int num_elements = bit_mask::num_elements(num_rows);
+  int block_size = 256;
+  int grid_size = (num_elements + block_size - 1)/block_size;
+
+  uint32_t *local_valid = (uint32_t *) malloc(num_elements * sizeof(uint32_t));
+  for (int i = 0 ; i < num_elements ; ++i) {
+    local_valid[i] = 0x55555555U;
+  }
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::copy_bit_mask(bit_mask.get_valid(), local_valid, num_rows, cudaMemcpyHostToDevice));
+
+  auto start = std::chrono::system_clock::now();
+  cudaProfilerStart();
+  for(int i = 0; i < 1000; ++i) {
+    gdf_size_type local_count = 0;
+    count_bits(&local_count, bit_mask, grid_size, block_size);
+  }
+  cudaProfilerStop();
+  auto end = std::chrono::system_clock::now();
+  std::chrono::duration<double> elapsed_seconds = end-start;
+  std::cout << "Elapsed time (ms): " << elapsed_seconds.count()*1000 << std::endl;
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+  free(local_valid);
+}
+
+TEST_F(BitMaskTest, MultiThreadedTest)
```

Comment:
"Multithreaded" implies CPU threads. This is a CUDA test. I would rename this to "CudaThreadingTest". I also think you should add a CPU multi-threading test.

Reply:
Not sure what we would want a CPU multi-threading test to do?

I don't implement a thread-safe version of the set/clear methods on the CPU.  I provided the methods that can be implemented on the CPU to facilitate unit testing (a bunch of the existing unit tests do single-threaded things on the CPU).  My impression was that this would be a data structure used on the GPU.

Reply:
@ChuckHastings your assumption is correct. I think it's sufficient to rename the test. There's no need for a CPU multi-threading test as this is not a supported or desired use-case.

Reply:
Sorry, I didn't notice that this is all __device__ code.